kratos/utilities/qef_utility.cpp

```diff
@@ -0,0 +1,115 @@
+        array_1d<double,3> center = CalculateCenter(rVoxel);
+        MatrixType m_center(3,1);
+        column(m_center,0) = center;
+        GeometryArrayType edges = rVoxel.GenerateEdges();
+
+        //Initialize the corresponding matrixes
+        MatrixType ata(3,3,0);  //3x3 matrix initialized to 0
+        MatrixType atb(3,1,0);  //3x1 matrix
+        MatrixType btb(1,1,0);  //1x1 matrix
+
+        for (std::size_t i = 0; i < rTriangles.size(); i++) {
+            array_1d<double,3> normal = CalculateNormal(rTriangles[i]);
+            MatrixType m_normal(3,1);
+            column(m_normal,0) = normal;
+
+            //We will iterate through the edges using a while loop, so that if a triangles intersects 2 edges (unlikely 
+            //but possible), only one will be taken into account to create the matrixes.
+            int result = 0; 
+            array_1d<double,3> intersection;
+            std::size_t j = 0;
+            while(!result && j < edges.size()) { 
+                PointsArrayType ends = edges[j++].Points();
+                result = IntersectionUtilities::ComputeTriangleLineIntersection(rTriangles[i],ends[0],ends[1],intersection);
+            }
+            if (result) {
+                //Fill the matrixes with the corresponding information from the intersection and normal
+                MatrixType m_normal_trans = trans(m_normal);
+                ata = ata + prod(m_normal,m_normal_trans);
+                double aux = MathUtils<double>::Dot(normal,intersection);
+                MatrixType m_aux(1,1,aux);
+                atb = atb + prod(m_normal,m_aux);
+                btb = btb + prod(m_aux,m_aux); 
+            }
+        }
+
+        //Find the eigenvalues and eigenvectors to ata
+        MatrixType m_eigenvectors;
+        MatrixType m_eigenvalues;
+        const bool converged = MathUtils<double>::GaussSeidelEigenSystem(ata, m_eigenvectors, m_eigenvalues);
+
+        MatrixType d(3,3,0);
+        for (int i : {0,1,2}) {
+            if (m_eigenvalues(i,i) < 1e-12) {
+                d(i,i) = 0; 
+            } else {
+                d(i,i) = Check(1.0/m_eigenvalues(i,i), 1e-12);
+            }
+        }
+
+        MatrixType ata_inverse;  
+        MathUtils<double>::BDBtProductOperation(ata_inverse, d, m_eigenvectors);
+       
+        MatrixType ata_c = prod(ata,m_center);
+        MatrixType solution = prod(ata_inverse, atb - ata_c) + m_center;        
+        x_point = column(solution,0);
+
+        return x_point;
+    }
+
+
+    array_1d<double,3> QEF::CalculateCenter(const GeometryType& rVoxel) {
+        PointsArrayType nodes = rVoxel.Points();
+        double x, y, z;
+        x = y = z = 0;
+        for(std::size_t i = 0; i < nodes.size(); i++) {
+            x += nodes[i].X();
+            y += nodes[i].Y();
+            z += nodes[i].Z();
+        }
+        array_1d<double,3> center({x/nodes.size(),y/nodes.size(),z/nodes.size()});
+        return center;
+    }
```

Comment:
Geometry already provides a `Center` method.

Reply:
So you can remove this and use that one.

Reply:
Done thanks!